typescript/schedules.md

```diff
@@ -0,0 +1,1255 @@
+
+const scheduleDescription = await schedule.describe();
+
+// For later, pending API finalization:
+// https://github.com/temporalio/proposals/pull/62/files#r933532170
+// const matchingStartTimes = await schedule.listMatchingTimes({ start: new Date(), end: new Date() })
+// const matchingStartTimes = await client.listMatchingTimes({ spec, start, end })
+
+await schedule.update(
+  (schedule) => {
+    schedule.spec.intervals[0].every = '1d';
+    // unset with:
+    // delete schedule.spec.intervals;
+
+    // return false to stop retrying early
+  },
+  { retry: retryPolicy }
+);
+
+await schedule.trigger();
+await schedule.backfill({ startAt: new Date(), endAt: addWeeks(new Date(), 1) }); // also takes array
+await schedule.pause('note: pausing');
+await schedule.unpause('now unpause');
+
+await schedule.delete();
+
+const { schedules, nextPageToken } = await client.listByPage({
+  pageSize: 50,
+  nextPageToken: 'base64',
+});
+
+for await (const schedule: Schedule of client.list()) {
+  const { id, memo, searchAttributes } = schedule;
+  // ...
+}
+```
+
+### Types
+
+```ts
+import { DataConverter, LoadedDataConverter } from '@temporalio/common';
+import { loadDataConverter } from '@temporalio/internal-non-workflow-common';
+import {
+  composeInterceptors,
+  Headers,
+  Replace,
+  RetryPolicy,
+  SearchAttributes,
+  Workflow,
+} from '@temporalio/internal-workflow-common';
+import { temporal } from '@temporalio/proto';
+import os from 'os';
+import { Connection } from './connection';
+import { ScheduleClientCallsInterceptor, ScheduleClientInterceptors } from './interceptors';
+import { ConnectionLike, Metadata, WorkflowService } from './types';
+import { WorkflowHandle, WorkflowStartOptions } from './workflow-client';
+
+import type { NonNegativeInteger, RequireAtLeastOne } from 'type-fest';
+
+// TODO is a non-generic NonNegativeInteger possible? The ones below error due to no type argument
+export type PositiveInteger = Exclude<NonNegativeInteger, 0>;
+
+/**
+ * @format number of milliseconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Milliseconds = string | NonNegativeInteger;
+
+/**
+ * @format number of seconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Seconds = string | NonNegativeInteger;
+
+export interface UpdateScheduleOptions {
+  /**
+   * @default TODO
+   */
+  retry?: RetryPolicy;
+}
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  startAt: Date;
+  endAt: Date;
+
+  /** Override Overlap Policy for this request. */
+  overlapPolicyOverride?: ScheduleOverlapPolicy;
+}
+
+/**
+ * Handle to a single Schedule
+ */
+export interface ScheduleHandle {
+  /**
+   * This Schedule's identifier
+   */
+  readonly id: string;
+
+  /**
+   * Update the Schedule
+   */
+  update(mutateFn: (schedule: ScheduleDescription) => void | false, options?: UpdateScheduleOptions): Promise<void>;
```

Comment:
So the update function can only return void or false, not true?

Reply:
Yes, I should be able to `return shouldUpdate` and not have my compiler break. Also, taking void + anything in a single-expression lambda can be confusing to a user, e.g. `desc => doUpdate(desc, { someOption: 'foo' })`.

Whether we want to ever accept void is subject to discussion. One could argue that a default return value (or throw) does not update. Void is often undefined which is falsey in most devs eyes. Maybe we should always require an explicit return of bool? And if the JS guys forget their return and have no type checking, at least their lack of return will mean no update.

Reply:
> Yes, I should be able to return shouldUpdate where shouldUpdate is a bool and not have my compiler break.

makes sense, thanks.

> Whether we want to ever accept void

my guess is the most common usage will never return false, in which case it's easier/less code to not return than to have to add `return true`.

an alternative that's closer to our Activity Retry Policy throwing nonRetryable: return type `void` and `throw StopRetrying()`

> Also, taking void + anything in a single-expression lambda can be confusing to a user, e.g. desc => doUpdate(desc, { someOption: 'foo' }).

what's confusing here?

Reply:
I do agree that most of the time you'd want to apply the update and rarely there'd be no update required.
One way to achieve this is to use something like immerjs to detect if anything changed instead of requiring the user to return a value.
Other options include, user returning the modified object or undefined or throwing a specific exception to indicate no update is required.
I like the immerjs approach but the cost would be an extra dependency.

Reply:
> what's confusing here?

That now I have to type my `doUpdate` as `void|false` instead of just `bool`.

I think returning a bool is clearest (e.g. I might not want to update it if it's at a certain state), but if y'all really think I need to use an exception then ok. I am not a fan of optional return types though as it's not clear to a user when they write a function whether they should return or not.

Reply:
I prefer immer

Reply:
I definitely think I can make the the Python and Go SDK API for all of this to be simpler and more straightforward. Merging is scary, adding dependencies is silly, etc. So long as all of this stuff we do in TypeScript doesn't infect our other SDKs, no prob.

Reply:
At this point my vote would be for raising a **special** exception and avoiding the extra dependencies.
I think in the common case you'd want to do the update so the exception would rarely need to be used.
I wouldn't treat any exception as opting out of the update because that could hide problems in the update handler, other exceptions should be propagated to the update method caller.
The immer approach is nice but I prefer being explicit with the exception route over the implicitness of immer (which also adds another dependency).

Other SDKs are free to implement this as they please.

Reply:
> Other SDKs are free to implement this as they please.

and TS: void & throw